core/src/main/java/org/elasticsearch/cluster/service/TaskBatching.java

```diff
@@ -0,0 +1,233 @@
+                        }
+                    }
+                }
+            }
+
+            if (toExecute.isEmpty() == false) {
+                final String tasksSummary = processTasksBySource.entrySet().stream().map(entry -> {
+                    String tasks = updateTask.describeTasks(entry.getValue());
+                    return tasks.isEmpty() ? entry.getKey() : entry.getKey() + "[" + tasks + "]";
+                }).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
+
+                run(updateTask.batchingKey, toExecute, tasksSummary);
+            }
+        }
+    }
+
+    /**
+     * Action to be implemented by the specific batching implementation
+     */
+    protected abstract void run(Object batchingKey, List<? extends BatchingTask> tasks, String tasksSummary);
+
+    /**
+     * Represents a runnable task that supports batching.
+     * Implementors of TaskBatching can subclass this to add a payload to the task.
+     */
+    protected abstract class BatchingTask extends SourcePrioritizedRunnable {
+        /**
+         * whether the task has been processed already
+         */
+        protected final AtomicBoolean processed = new AtomicBoolean();
+
+        /**
+         * the object that is used as batching key
+         */
+        protected final BatchingKey<?> batchingKey;
+        /**
+         * the task object that is wrapped
+         */
+        protected final Object taskIdentity;
+
+        protected BatchingTask(Priority priority, String source, BatchingKey<?> batchingKey, Object taskIdentity) {
+            super(priority, source);
+            this.batchingKey = batchingKey;
+            this.taskIdentity = taskIdentity;
+        }
+
+        @Override
+        public void run() {
+            runIfNotProcessed(this);
+        }
+
+        @Override
+        public String toString() {
+            String taskDescription = describeTasks(Collections.singletonList(this));
+            if (taskDescription.isEmpty()) {
+                return "[" + source + "]";
+            } else {
+                return "[" + source + "[" + taskDescription + "]]";
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        public String describeTasks(List<? extends BatchingTask> tasks) {
+            return ((BatchingKey) batchingKey).describeTasks(
+                tasks.stream().map(BatchingTask::getTaskIdentity).collect(Collectors.toList()));
+        }
+
+        public Object getTaskIdentity() {
+            return taskIdentity;
+        }
+    }
+
+    public interface BatchingKey<T> {
+        /**
+         * Builds a concise description of a list of tasks (to be used in logging etc.).
+         *
+         * This method can be called multiple times with different lists before execution.
+         * This allows groupd task description but the submitting source.
+         */
+        default String describeTasks(List<T> tasks) {
```

Comment:
I'm not sure we need this interface. It feels weird to me that the batchingKey knows how to describe the tasks. I would prefer to have an abstract method `describeTasks(List<? extends BatchedTask>, batchKey)` which can be overridden by the subclasses. This will be consistent with run and onTimeout (see other comment) and will allows us to remove this interface.